### PR TITLE
[flang][StackArrays] skip analysis of very large functions

### PR DESCRIPTION
**Pull request by tblah**
The stack arrays pass uses data flow analysis to determine whether heap allocations are freed on all paths out of the function.

`interp_domain_em_part2` in spec2017 wrf generates over 120k operations, including almost 5k fir.if operations and over 200 fir.do_loop operations, all in the same function. The MLIR data flow analysis framework cannot provide reasonable performance for such cases because there is a combinatorial explosion in the number of control flow paths through the function, all of which must be checked to determine if the heap allocations will be freed.

This patch skips the stack arrays pass for ridiculously large functions (defined as having more than 1000 fir.allocmem operations). This threshold is configurable at runtime with a command line argument.

With this patch, compiling this file is more than 80% faster.

### CODE REVIEW

**Comment by tblah** [comment]
@d-smirnov for some reason it didn't let me add you as a reviewer

File: flang/lib/Optimizer/Transforms/StackArrays.cpp [hunk 1]
```diff
@@ -411,6 +417,17 @@ void AllocationAnalysis::processOperation(mlir::Operation *op) {
 mlir::LogicalResult
 StackArraysAnalysisWrapper::analyseFunction(mlir::Operation *func) {
   assert(mlir::isa<mlir::func::FuncOp>(func));
+  size_t nAllocs = 0;
+  func->walk([&nAllocs](fir::AllocMemOp) { nAllocs++; });
+  // don't bother with the analysis if there are no heap allocations
+  if (nAllocs == 0)
+    return mlir::success();
+  if ((maxAllocsPerFunc != 0) && (nAllocs > maxAllocsPerFunc)) {
```
[the remark this answers is not in the document]
Comment (tblah) [reply flang/lib/Optimizer/Transforms/StackArrays.cpp:425]: I was intending this to follow the idiom of "set the limit to zero for unlimited"

[the remark this answers is not in the document]
Comment (kiranchandramohan) [reply flang/lib/Optimizer/Transforms/StackArrays.cpp:425]: I see. May be just document that.

**Review by kiranchandramohan — APPROVED** [review]
LG with comment added.

**Pull request merged**